packages/@lwc/ssr-compiler/src/estemplate.ts

```diff
@@ -12,45 +12,98 @@ import type {
     Node as EsNode,
     Program as EsProgram,
     FunctionDeclaration as EsFunctionDeclaration,
+    Statement as EsStatement,
 } from 'estree';
+import type { Checker } from 'estree-toolkit/dist/generated/is-type';
 
-export const placeholder = false;
-type ReplacementNode = EsNode | EsNode[] | null;
+/** Placeholder value to use to opt out of validation. */
+const NO_VALIDATION = false;
 
-type ReturnsBool = (node: EsNode | null) => boolean;
-export type Validator = ReturnsBool | typeof placeholder;
+/** A function that accepts a node and checks that it is a particular type of node. */
+type Validator<T extends EsNode | null = EsNode | null> = (
+    node: EsNode | null | undefined
+) => node is T;
+
+/**
+ * A pointer to a previous value in the template literal, indicating that the value should be re-used.
+ * @see {@linkcode esTemplate}
+ */
+type ValidatorReference = number;
+
+/** A validator, validation opt-out, or reference to previously-used validator. */
+type ValidatorPlaceholder<T extends EsNode | null> =
+    | Validator<T>
+    | ValidatorReference
+    | typeof NO_VALIDATION;
+
+/**
+ * Converts the validators and refs used in the template to the list of paramters required by the
+ * created template function. Removes back references to previous slots from the list.
+ */
+type ToReplacementParameters<Arr extends unknown[]> = Arr extends [infer Head, ...infer Rest]
+    ? // `Checker` is estree's validator type, and has overloads that mess with inferencing
+      Head extends Checker<infer C>
+        ? [C | C[], ...ToReplacementParameters<Rest>]
+        : Head extends Validator<infer V>
+          ? // If it's a validator, extract the validated type
+            null extends V
+              ? // Avoid `(V | null)[]` if it's nullable
+                [V | (V & {})[], ...ToReplacementParameters<Rest>]
+              : [V | V[], ...ToReplacementParameters<Rest>]
+          : Head extends typeof NO_VALIDATION
+            ? // If it's NO_VALIDATION, use `unknown`
+              [unknown, ...ToReplacementParameters<Rest>]
+            : // If it's a back ref, just drop it
+              ToReplacementParameters<Rest>
+    : [];
```

Comment:
This type does a lot of heavy lifting! To perform array operations (map/filter/etc.) in TypeScript, you have to resort to recursion. The general pattern looks like this:
```ts
// T -> Whatever
type ChangeValue<T> = Whatever

// T[] -> Whatever[]
type ChangeArray<Arr extends unknown[]> = Arr extends [infer Head, ...infer Rest]
  // If the array has one or more items, change the first, then recursively change the rest
  ? [ChangeValue<Head>, ...ChangeArray<Rest>]
  // The array is already empty -- don't need to do anything
  : []

// Filter values not assignable to `BaseType`
type FilterArray<Arr extends unknown[], BaseType> = Arr extends [infer Head, ...infer Rest]
  ? Head extends Basetype
    ? [Head, ...FilterArray<Rest>] // keep the value
    : FilterArray<Rest> // ditch the value
  : [] // array is already empty
```

Within that pattern, we use the `infer` keyword to extract the validated types from the validators. For example:
```ts
type PromiseValue<P extends Promise<any>> = P extends Promise<infer T> ? T : never
type PromisedAbc = Promise<'abc'>
type Abc = PromiseValue<PromisedAbc> // 'abc'
```

Sprinkle a few edge cases into the mix (as described in code comments), and you get this delightful monstrosity!